### PR TITLE
enforce the interface relationship between ServicePort and BaseServiceInfo

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
Currently the BaseServiceInfo struct implements the ServicePort interface, but
only uses that interface sometimes. All the elements of BaseServiceInfo are exported
and sometimes the interface is used to access them and othertimes not

I extended the ServicePort interface so that all relevent values can be accessed through
it and unexported all the elements of BaseServiceInfo

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
>
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/types.go [hunk 1]
```diff
@@ -48,8 +49,14 @@ func (spn ServicePortName) String() string {
 type ServicePort interface {
 	// String returns service string.  An example format can be: `IP:Port/Protocol`.
 	String() string
-	// ClusterIPString returns service cluster IP in string format.
-	ClusterIPString() string
+	// GetClusterIP returns service cluster IP in net.IP format.
+	GetClusterIP() net.IP
```
Comment (danwinship) [inline pkg/proxy/types.go]: golang style is not to use "Get" in getter names (https://golang.org/doc/effective_go.html#Getters)

**Review by danwinship — COMMENTED** [review]
What inspired this? Is this a prologue to some other change?

File: pkg/proxy/types.go (hunk 1 above)
Comment (JacobTanenbaum) [reply pkg/proxy/types.go]: I used "Get" in the getter name because some of them already did and now a majority of them do now. Should I remove "Get" from all of them?

I changed from string to net.IP because sometimes the string version of the IP address and other times they just accessed the net.IP through BaseServiceInfo.ClusterIP. I thought instead of having two getters since net.IP has a string method it would be better to return the more generic type and call the string function when needed

**Comment by JacobTanenbaum** [comment]
@danwinship I found it confusing the last time I changed kubeproxy that the interface existed but wasn't used consistently

File: pkg/proxy/types.go (hunk 1 above)
Comment (danwinship) [reply pkg/proxy/types.go]: Hm... it appears that the interface used to have `ClusterIP()`, `Protocol()`, and `HealthCheckNodePort()`, and @MrHohn changed it in https://github.com/kubernetes/kubernetes/commit/b485f7b5 as part of #56880. (Renaming the methods to have `Get` so as to not conflict with newly-added public struct fields named `ClusterIP`, `Protocol`, and `HealthCheckNodePort`.) I'm not sure if there's any good reason for it...

Anyway, if you're making the struct fields private and consistently using the interface methods, then you should make the interface method names not have "Get" in them.

**Comment by danwinship** [comment]
Drop pkg/scheduler/apis/config/v1alpha1/zz_generated.conversion.go from the patch. (Looks like you did an update/regen with a slightly different version of tools and got irrelevant changes.)

File: pkg/proxy/types.go (hunk 1 above)
Comment (MrHohn) [reply pkg/proxy/types.go]: Thanks for adding me. @danwinship is correct, I made these this way purely to avoid conflicting. +1 to not having the redundant `Get` if no longer needed.

**Comment by JacobTanenbaum** [comment]
@dcbw @danwinship PTAL

**Comment by dcbw** [comment]
@freehan @thockin you guys able to approve this one? Thanks!

**Comment by JacobTanenbaum** [comment]
@freehan @thockin does this need more work?

File: pkg/proxy/iptables/proxier_test.go [hunk 2]
```diff
@@ -993,7 +992,6 @@ func onlyLocalNodePorts(t *testing.T, fp *Proxier, ipt *iptablestest.FakeIPTable
 	fp.nodePortAddresses = []string{"10.20.30.0/24"}
 
 	fp.syncProxyRules()
-
```
Comment (dcbw) [inline pkg/proxy/iptables/proxier_test.go]: unnecessary whitespace removal?

File: pkg/proxy/iptables/proxier_test.go [hunk 3]
```diff
@@ -276,18 +249,44 @@ func TestDeleteEndpointConnections(t *testing.T) {
...
+		input := []proxy.ServiceEndpoint{
+			{
+				Endpoint:        tc.endpoint,
+				ServicePortName: svc,
+			},
+		}
+
+		fp.deleteEndpointConnections(input)
 
-		// For UDP connections, check the executed conntrack command
```
Comment (dcbw) [inline pkg/proxy/iptables/proxier_test.go]: Did you mean to remove this comment?
